flowapi/flowapi/permissions.py

```diff
@@ -1,259 +1,72 @@
-    ----------
-    x : tuple
-        Item to get a key for.
-
-    Returns
-    -------
-    tuple of bool, int and input or tuple of bool and input
-        False, negative length of input and then key for dicts and lists
-        True, key for other tuples
-
-    """
-    if isinstance(x[1], (dict, list)):
-        return False, -len(x[1]), x[0]
-    else:
-        return True, x[0]
-
-
-@functools.singledispatch
-def valid_tree_walks(
-    tree,
-    *,
-    paths: Optional[Iterable[str]] = None,
-    depth: int = 1,
-) -> Union[tuple, list]:
-    """
-
-    Parameters
-    ----------
-    tree : list of str or dict
-        Tree to walk through
-    paths : iterable of str or None
-        Nodes already traversed
-    depth : int, default 1
-        Current depth in the tree
-
-
-    Yields
-    ------
-    list of str or tuple of list of str
-        Path through the tree
-
-    """
-    yield [*paths, tree]
+from quart import request, current_app
 
 
-@valid_tree_walks.register
-def _(
-    tree: list,
-    *,
-    paths: Optional[Iterable[str]] = None,
-    depth: int = 1,
-) -> Union[tuple, list]:
-    if paths is None:
-        paths = tuple()
-    if len(tree) == 0 and len(paths) > 0:
-        yield list(paths)
-    else:
-        for v in tree:
-            yield [*paths, v]
+from flowapi.flowapi_errors import (
+    JSONHTTPException,
+    MissingQueryKindError,
+    BadQueryError,
+)
 
+try:
+    logger = current_app.debug_logger
+except RuntimeError:
+    import structlog
 
-@valid_tree_walks.register
-def _(
-    tree: dict,
-    *,
-    paths: Optional[Iterable[str]] = None,
-    depth: int = 1,
-) -> Union[tuple, list]:
-    even_depth = depth % 2 == 0
-    if paths is None:
-        paths = tuple()
-    if len(tree) == 0 and len(paths) > 0:
-        yield list(paths)
-    elif not even_depth or len(tree) == 1:
-        for k, v in sorted(tree.items(), key=dict_and_list_key_function):
-            yield from valid_tree_walks(v, paths=(*paths, k), depth=depth + 1)
-    else:
-        yield from product(
-            *(
-                valid_tree_walks(v, paths=(*paths, k), depth=depth + 1)
-                for k, v in sorted(
-                    tree.items(),
-                    key=dict_and_list_key_function,
-                )
-            )
-        )
+    logger = structlog.get_logger(__name__)
 
 
 @functools.singledispatch
-def tree_walk_to_scope_list(tree_walk) -> str:
-    """
-
-    Parameters
-    ----------
-    tree_walk : str, list or str or tuple of list of str
-        A path through the tree
-
-    Yields
-    ------
-    str
-        The tree walk as a . delimited string
-
-    """
-    yield tree_walk
-
-
-@tree_walk_to_scope_list.register
-def _(tree_walk: list) -> str:
-    yield ".".join(tree_walk)
-
-
-@tree_walk_to_scope_list.register
-def _(tree_walk: tuple) -> str:
-    for walk in tree_walk:
-        yield from tree_walk_to_scope_list(walk)
-
+def grab_on_key_list(in_iter, search_keys):
```

Comment:
I think this func could really do with a doctoring, because it is potentially pretty difficult to follow